alembic/versions/0021_added_created_and_last_modified_columns_.py

```diff
@@ -0,0 +1,372 @@
+            server_default=sa.text("now()"),
+            nullable=False,
+        ),
+    )
+    op.add_column(
+        "family",
+        sa.Column(
+            "created",
+            sa.DateTime(timezone=True),
+            server_default=sa.text("now()"),
+            nullable=False,
+        ),
+    )
+    op.add_column(
+        "family",
+        sa.Column(
+            "last_modified",
+            sa.DateTime(timezone=True),
+            server_default=sa.text("now()"),
+            nullable=False,
+        ),
+    )
+    op.add_column(
+        "family_event",
+        sa.Column(
+            "created",
+            sa.DateTime(timezone=True),
+            server_default=sa.text("now()"),
+            nullable=False,
+        ),
+    )
+    op.add_column(
+        "family_event",
+        sa.Column(
+            "last_modified",
+            sa.DateTime(timezone=True),
+            server_default=sa.text("now()"),
+            nullable=False,
+        ),
+    )
+    public_update_1_last_modified = PGFunction(
+        schema="public",
+        signature="update_1_last_modified()",
+        definition="""
+    RETURNS TRIGGER AS $$
+    BEGIN
+        NEW.last_modified = NOW();
+        RETURN NEW;
+    END;
+    $$ language 'plpgsql'""",
+    )
+    op.create_entity(public_update_1_last_modified)  # type: ignore
+
+    public_update_2_family_last_modified = PGFunction(
+        schema="public",
+        signature="update_2_family_last_modified()",
+        definition="""
+    RETURNS TRIGGER AS $$
+    BEGIN
+        UPDATE family
+        SET last_modified = NOW()
+        WHERE import_id = NEW.family_import_id;
+        RETURN NEW;
+    END;
+    $$ language 'plpgsql'""",
+    )
+    op.create_entity(public_update_2_family_last_modified)  # type: ignore
+
+    public_update_2_collection_last_modified = PGFunction(
+        schema="public",
+        signature="update_2_collection_last_modified()",
+        definition="""
+    RETURNS TRIGGER AS $$
+    BEGIN
+        UPDATE collection
+        SET last_modified = NOW()
+        WHERE import_id = NEW.collection_import_id;
+        RETURN NEW;
+    END;
+    $$ language 'plpgsql'""",
+    )
+    op.create_entity(public_update_2_collection_last_modified)  # type: ignore
+
+    public_family_document_update_last_modified = PGTrigger(
+        schema="public",
+        signature="update_last_modified",
+        on_entity="public.family_document",
+        is_constraint=False,
+        definition="""
+    BEFORE UPDATE ON public.family_document
+    FOR EACH ROW
+    EXECUTE PROCEDURE public.update_1_last_modified()""",
+    )
+    op.replace_entity(public_family_document_update_last_modified)  # type: ignore
+
+    public_family_event_update_last_modified = PGTrigger(
+        schema="public",
+        signature="update_last_modified",
+        on_entity="public.family_event",
+        is_constraint=False,
+        definition="""
+    BEFORE UPDATE ON public.family_event
```

Comment:
Ah, it might seem like it but actually no! The inserts are taken care of by default DDL values. Some things we soft delete (which to the DB is an update statement) and some things we hard delete. We cannot update the last modified time of a family event if the record is deleted from the family_event table, so we instead update the last_modified on family when this happens. It confused the hell out of me for most of this week trying to get the expected behaviour to work! :confused:

Reply:
Ohhh interesting